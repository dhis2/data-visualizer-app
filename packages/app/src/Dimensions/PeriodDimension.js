import React, { Component } from 'react';
import PeriodSelectorDialog from '@dhis2/d2-ui-period-selector-dialog';
import i18n from '@dhis2/d2-i18n';

export class PeriodDimension extends Component {
    render = () => {
        return (
            <PeriodSelectorDialog
                open={false}
                onClose={() => console.log('closing period')}
                onUpdate={() =>
                    console.log('Period selector received argument')
                }
                d2={console.log('Period selector missing argument')}
            />
        );
    };
}

export default PeriodDimension;
