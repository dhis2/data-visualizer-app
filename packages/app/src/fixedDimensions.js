export const FIXED_DIMENSIONS = {
    Data: {
        id: 'Data',
        displayName: 'Data',
        iconName: 'DataIcon',
    },
    Period: {
        id: 'Period',
        displayName: 'Period',
        iconName: 'PeriodIcon',
    },
    OrgUnit: {
        id: 'OrgUnit',
        displayName: 'Organisation Unit',
        iconName: 'OrgUnitIcon',
    },
};
