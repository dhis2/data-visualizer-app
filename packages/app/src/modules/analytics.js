export const computeGenericPeriodNames = responses => {
    const xAxisRes = responses.reduce((out, res) => {
        if (out.metaData) {
            if (
                res.metaData.dimensions.pe.length >
                out.metaData.dimensions.pe.length
            ) {
                out = res;
            }
        } else {
            out = res;
        }

        return out;
    }, {});

    const metadata = xAxisRes.metaData;

    return metadata.dimensions.pe.reduce((genericPeriodNames, periodId) => {
        const name = metadata.items[periodId].name;

        // until the day the backend will support this in the API:
        // trim off the trailing year in the period name
        // english names should all have the year at the end of the string
        genericPeriodNames.push(name.replace(/\s+\d{4}$/, ''));

        return genericPeriodNames;
    }, []);
};
