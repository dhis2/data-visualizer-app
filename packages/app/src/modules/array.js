export const toArray = param => (Array.isArray(param) ? param : [param]);
