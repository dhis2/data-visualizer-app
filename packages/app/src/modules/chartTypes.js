import i18n from '@dhis2/d2-i18n';

export const COLUMN = 'COLUMN';
export const STACKED_COLUMN = 'STACKED_COLUMN';
export const BAR = 'BAR';
export const STACKED_BAR = 'STACKED_BAR';
export const LINE = 'LINE';
export const AREA = 'AREA';
export const PIE = 'PIE';
export const RADAR = 'RADAR';
export const GAUGE = 'GAUGE';
export const BUBBLE = 'BUBBLE';
export const YEAR_ON_YEAR = 'YEAR_ON_YEAR';

export const chartTypeDisplayNames = {
    [COLUMN]: i18n.t('Column'),
    [STACKED_COLUMN]: i18n.t('Stacked column'),
    [BAR]: i18n.t('Bar'),
    [STACKED_BAR]: i18n.t('Stacked bar'),
    [LINE]: i18n.t('Line'),
    [AREA]: i18n.t('Area'),
    [PIE]: i18n.t('Pie'),
    [RADAR]: i18n.t('Radar'),
    [GAUGE]: i18n.t('Gauge'),
    [YEAR_ON_YEAR]: i18n.t('Year on year'),
};
