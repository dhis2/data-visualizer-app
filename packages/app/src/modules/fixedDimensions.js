import i18n from '@dhis2/d2-i18n';
import DataIcon from '../assets/DataIcon';
import PeriodIcon from '../assets/PeriodIcon';
import OrgUnitIcon from '../assets/OrgUnitIcon';

export const FIXED_DIMENSIONS = {
    dx: {
        id: 'dx',
        name: i18n.t('Data'),
        iconName: 'DataIcon',
        icon: DataIcon,
    },
    pe: {
        id: 'pe',
        name: i18n.t('Period'),
        iconName: 'PeriodIcon',
        icon: PeriodIcon,
    },
    ou: {
        id: 'ou',
        name: i18n.t('Organisation Unit'),
        iconName: 'OrgUnitIcon',
        icon: OrgUnitIcon,
    },
};
