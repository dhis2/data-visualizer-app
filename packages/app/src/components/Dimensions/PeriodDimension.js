import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DialogContent from '@material-ui/core/DialogContent';
import DialogActions from '@material-ui/core/DialogActions';

import { PeriodSelector } from '@dhis2/d2-ui-period-selector-dialog';
import i18n from '@dhis2/d2-i18n';

import { sGetUi } from '../../reducers/ui';
import { sGetMetadata } from '../../reducers/metadata';
import { acSetCurrentFromUi } from '../../actions/current';
import { acRemoveUiItems, acAddUiItems } from '../../actions/ui';
import { acAddMetadata } from '../../actions/metadata';

import { HideButton, UpdateButton } from './DataDimension/buttons';
import { styles } from './styles/PeriodDimension.style';

const PE = 'pe';
const PERIOD = 'PERIOD';

export class PeriodDimension extends Component {
    onUpdateClick = () => {
        this.props.onUpdate(this.props.ui);
        this.props.toggleDialog(null);
    };

    selectPeriodDimensions = periods => {
        const idsToAdd = periods.map(periodRange => periodRange.id);

        this.props.addUiItems({
            dimensionType: PE,
            value: idsToAdd,
        });

        const arrToId = periods.reduce((obj, item) => {
            obj[item.id] = { ...item, dimensionItemType: PERIOD };
            return obj;
        }, {});

        this.props.addMetadata(arrToId);
    };

    deselectPeriodDimensions = periods => {
        const idsToRemove = periods.map(periodRange => periodRange.id);

        this.props.removeUiItems({
            dimensionType: PE,
            value: idsToRemove,
        });
    };

    getSelectedPeriods = () => {
        return this.props.ui.itemsByDimension[PE].map(item => ({
            id: item,
            name: this.props.metadata[item].name,
        }));
    };

    render = () => {
        const selectedPeriods = this.getSelectedPeriods();

        return (
            <div style={styles.container}>
                <DialogContent style={styles.dialogContent}>
                    <h3 style={styles.dialogTitle}>{i18n.t('Period')}</h3>
                    <PeriodSelector
                        d2={this.context.d2}
                        onSelect={this.selectPeriodDimensions}
                        onDeselect={this.deselectPeriodDimensions}
                        selectedItems={selectedPeriods}
                    />
                </DialogContent>
                <DialogActions style={styles.dialogActions}>
                    <HideButton action={() => this.props.toggleDialog(null)} />
                    <UpdateButton action={this.onUpdateClick} />
                </DialogActions>
            </div>
        );
    };
}

const mapStateToProps = state => ({
    metadata: sGetMetadata(state),
    ui: sGetUi(state),
});

export default connect(
    mapStateToProps,
    {
        addMetadata: acAddMetadata,
        addUiItems: acAddUiItems,
        removeUiItems: acRemoveUiItems,
        onUpdate: acSetCurrentFromUi,
    }
)(PeriodDimension);

PeriodDimension.propTypes = {
    toggleDialog: PropTypes.func.isRequired,
    ui: PropTypes.object.isRequired,
};

PeriodDimension.contextTypes = {
    d2: PropTypes.object,
};
