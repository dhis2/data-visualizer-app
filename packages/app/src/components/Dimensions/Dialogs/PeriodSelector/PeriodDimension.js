import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import DialogTitle from '@material-ui/core/DialogTitle';
import DialogContent from '@material-ui/core/DialogContent';

import { PeriodSelector } from '@dhis2/d2-ui-period-selector-dialog';
import i18n from '@dhis2/d2-i18n';

import { sGetUi } from '../../../../reducers/ui';
import { sGetMetadata } from '../../../../reducers/metadata';
import { acRemoveUiItems, acAddUiItems } from '../../../../actions/ui';
import { acAddMetadata } from '../../../../actions/metadata';
import { FIXED_DIMENSIONS } from '../../../../modules/fixedDimensions';

const peId = FIXED_DIMENSIONS.pe.id;

const PERIOD = 'PERIOD';

export class PeriodDimension extends Component {
    selectPeriodDimensions = periods => {
        const idsToAdd = periods.map(periodRange => periodRange.id);

        this.props.addUiItems({
            dimensionType: peId,
            value: idsToAdd,
        });

        const arrToId = periods.reduce((obj, item) => {
            obj[item.id] = { ...item, dimensionItemType: PERIOD };
            return obj;
        }, {});

        this.props.addMetadata(arrToId);
    };

    deselectPeriodDimensions = periods => {
        const idsToRemove = periods.map(periodRange => periodRange.id);

        this.props.removeUiItems({
            dimensionType: peId,
            value: idsToRemove,
        });
    };

    getSelectedPeriods = () => {
        return this.props.ui.itemsByDimension[peId].map(item => ({
            id: item,
            name: this.props.metadata[item].name,
        }));
    };

    render = () => {
        const selectedPeriods = this.getSelectedPeriods();

        return (
            <Fragment>
                <DialogTitle> {i18n.t('Period')}</DialogTitle>
                <DialogContent>
                    <PeriodSelector
                        d2={this.context.d2}
                        onSelect={this.selectPeriodDimensions}
                        onDeselect={this.deselectPeriodDimensions}
                        selectedItems={selectedPeriods}
                    />
                </DialogContent>
            </Fragment>
        );
    };
}

PeriodDimension.propTypes = {
    ui: PropTypes.object.isRequired,
    metadata: PropTypes.object.isRequired,
    addMetadata: PropTypes.func.isRequired,
    addUiItems: PropTypes.func.isRequired,
    removeUiItems: PropTypes.func.isRequired,
};

PeriodDimension.contextTypes = {
    d2: PropTypes.object,
};

const mapStateToProps = state => ({
    metadata: sGetMetadata(state),
    ui: sGetUi(state),
});

export default connect(
    mapStateToProps,
    {
        addMetadata: acAddMetadata,
        addUiItems: acAddUiItems,
        removeUiItems: acRemoveUiItems,
    }
)(PeriodDimension);
