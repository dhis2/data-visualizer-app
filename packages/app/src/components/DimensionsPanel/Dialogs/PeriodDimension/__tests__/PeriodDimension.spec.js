import React from 'react';
import renderer from 'react-test-renderer';
import { PeriodDimension } from '../PeriodDimension';
import { FIXED_DIMENSIONS } from '../../../../../modules/fixedDimensions';

const peId = FIXED_DIMENSIONS.pe.id;

jest.mock('@dhis2/d2-ui-period-selector-dialog', () => ({
    PeriodSelector: 'mockPeriodSelector',
}));

describe('The Period Dimension component', () => {
    const props = {
        ui: {
            itemsByDimension: {
                [peId]: [],
            },
        },
        metadata: {},
        addMetadata: jest.fn(),
        addUiItems: jest.fn(),
        removeUiItems: jest.fn(),
        setUiItems: jest.fn(),
        context: { d2: {} },
    };

    it('renders correctly', () => {
        const tree = renderer.create(<PeriodDimension {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
