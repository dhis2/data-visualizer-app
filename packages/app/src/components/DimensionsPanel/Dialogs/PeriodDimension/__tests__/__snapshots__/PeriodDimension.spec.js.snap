// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Period Dimension component renders correctly 1`] = `
Array [
  <div
    className="MuiDialogTitle-root-1"
  >
    <h2
      className="MuiTypography-root-2 MuiTypography-title-8"
    >
       
      Period
    </h2>
  </div>,
  <div
    className="MuiDialogContent-root-38"
  >
    <mockPeriodSelector
      onDeselect={[Function]}
      onReorder={[Function]}
      onSelect={[Function]}
      selectedItems={Array []}
    />
  </div>,
]
`;
